main.go

```diff
@@ -456,16 +456,16 @@ func main() {
 			r.Use(authentication.WithTenant)
 			r.Use(authentication.WithTenantID(tenantIDs))
 
-			oidcHandler, oidcTenantMiddlewares, warnings := authentication.NewOIDC(logger, reg, "/oidc/{tenant}", oidcs)
-			for _, w := range warnings {
-				level.Warn(logger).Log("msg", w.Error())
+			oh := authentication.NewOIDCHandlers(logger, reg)
+			for _, oidc := range oidcs {
+				oh.AddOIDCForTenant("/oidc/{tenant}", oidc)
 			}
-			r.Mount("/oidc/{tenant}", oidcHandler)
+			r.Mount("/oidc/{tenant}", oh.Router())
 
 			// Metrics
 			if cfg.metrics.enabled {
 				r.Group(func(r chi.Router) {
-					r.Use(authentication.WithTenantMiddlewares(oidcTenantMiddlewares, authentication.NewMTLS(mTLSs)))
+					r.Use(authentication.WithTenantMiddlewares(oh.Middlewares(), authentication.NewMTLS(mTLSs)))
```

Comment:
This is only working right now because we are returning the map of middlewares, which is a reference. So even though the `Middlewares` func is only ever called ONCE, the map with the middlewares gets updated to include more tenant middlewares later. However, this hides a dangerous race: maps are not safe to access concurrently.

We need to change the `Middlewares` func to return a concurrent-safe object that can be accesses at the same time someone else who is locking the map is writing to it. This way we would normally do this would be to construct a new map that is a copy of the original `middlewares` map. However, because the `Middlewares` func is only called once EVER, it won't get updated.